[NVPTX] use incomplete aggregate initializers

AlexMaclean opened the pull request:
The PTX ISA specifies that initializers may be incomplete ([5.4.4. Initializers](https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#initializers))
>  As in C, array initializers may be incomplete, i.e., the number of initializer elements may be less than the extent of the corresponding array dimension, with remaining array locations initialized to the default value for the specified array type.

Emitting initializers in this form is preferable because it reduces the size of the PTX, in some cases significantly, and can improve compile time of ptxas as a result.

AlexMaclean committed:
[NVPTX] use incomplete aggregate initializers

Artem-B commented on llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp:
```diff
@@ -1292,10 +1292,20 @@ void NVPTXAsmPrinter::AggBuffer::printSymbol(unsigned nSym, raw_ostream &os) {
+  // Last contingous sequnce of 0 bytes are unintialized to get default
+  // initialization in ptxas. This avoids ptxas memory consumption for large
+  // aggregates.
+  unsigned int count = size;
```
This could use a more descriptive name. `InitializerCount` ?

AlexMaclean replied:
Fixed.

Artem-B commented on llvm/test/CodeGen/NVPTX/globals_init.ll:
```diff
@@ -26,6 +26,6 @@
 
 ; Make sure we fill in alignment gaps correctly.
-; CHECK-DAG: .b8 GblU[12] = {7, 6, 0, 0, 5, 4, 3, 2, 1, 0, 0, 0};
+; CHECK-DAG: .b8 GblU[12] = {7, 6, 0, 0, 5, 4, 3, 2, 1};
```
Please add few more cases:
- all-zeros,
- all-non-zeros
- initializer with a symbol

Also, please check with old versions of ptxas (10.2 is probably the oldest version we may still care about).

AlexMaclean replied:
All added in `llvm/test/CodeGen/NVPTX/global-incomplete-init.ll`. I've verified this works with 10.2 `ptxas`.

AlexMaclean committed:
address comments

AlexMaclean commented on llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp:
```diff
@@ -1292,10 +1292,20 @@ void NVPTXAsmPrinter::AggBuffer::printSymbol(unsigned nSym, raw_ostream &os) {
-  symbolPosInBuffer.push_back(size);
+  // Last contingous sequnce of 0 bytes are unintialized to get default
+  // initialization in ptxas. This avoids ptxas memory consumption for large
+  // aggregates.
```
Rephrased. 

>  Speaking of memory use, do you have any specific data on how much that affects ptxas?

Sorry, I don't have any specific data.

Artem-B replied:
Then, how do we know that it does avoid memory consumption w/ ptxas? If we do not, perhaps, drop the last sentence? Or replace it with "saves on generated ptx size" which is a good reason relevant to LLVM users.

AlexMaclean replied:
Fair enough. Here is some specific data: for a ptx file which defines a single 100,000 element `b8` array where the 0th element is 1 and the rest are 0s the maximum resident set size (as measured by `/usr/bin/time`) falls from 17,492 kbytes to 11,036 kybtes when the remainder of the default initialization is used.

AlexMaclean commented:
@Artem-B, I've addressed all comments, please let me know if there is anything else I should do to fix this up, if not please land on my behalf once tests pass.

Artem-B commented on llvm/lib/Target/NVPTX/NVPTXAsmPrinter.cpp:
```diff
@@ -1292,10 +1292,20 @@ void NVPTXAsmPrinter::AggBuffer::printSymbol(unsigned nSym, raw_ostream &os) {
+  // Do not emit trailing zero initializers. They will be zero-initialized by
+  // ptxas. This saves on both space requirements for the generated PTX and on
+  // memory use by ptxas.
+  unsigned int InitializerCount = size;
```
I'd also add a reference to the PTX spec: https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#global-state-space

Artem-B reviewed (approved):
LGTM with a comment nit.

AlexMaclean committed:
add docs link requested by review

AlexMaclean commented:
@Artem-B, could you please land this om my behalf?
